Lesson11/AdditionalTask1/Program.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Threading;
+
+namespace AdditionalTask1
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var thread = new ThreadWithResult<int>(() => 10);
+
+            Console.WriteLine($"Result: {thread.Result}");
+
+            thread.Start();
+
+            Console.WriteLine($"Result: {thread.Result}");
+
+            Console.ReadLine();
+        }
+    }
+
+    class ThreadWithResult<TResult>
+    {
+        private Thread _thread;
+        private TResult _result;
+
+        public ThreadWithResult(Func<TResult> action)
+        {
+            _thread = new Thread(() =>
+            {
+                try
+                {
+                    _result = action();
+                    IsSuccess = true;
+                }
+                catch (Exception ex)
+                {
+                    Exception = ex;
+                    IsSuccess = false;
+                }
+                finally
+                {
+                    IsCompleted = true;
+                }
+
+            });
+        }
+
+        public bool IsCompleted { get; private set; }
+        public bool IsStarted { get; private set; }
+        public bool IsSuccess { get; private set; }
+        public Exception Exception { get; private set; }
+        public TResult Result
+        {
+            get
+            {
+                if (IsCompleted)
+                {
+                    return IsSuccess ? _result : throw Exception;
+                }
+                else
+                {
+                    if (IsStarted)
+                    {
+                        Wait();
+                        return Result;
+                    }
+                    else
+                    {
+                        return default;
```

Comment:
It may result in not clear behavior. If you call `Result` on non-started instance of `ThreadWithResult` and it returns the default value (e.g `0` in your case) you cannot say if it is not started or it is finished with such result. So consider throwing exception or starting the thread and wait until it is completed

Reply:
Ok)